lavc : use external surface

popkartyeah opened the pull request:
1. use external surface fix the transcode block issue.
2. fix decode flush no frame issue.

Commit:
lavc : use external surface

1. use external surface fix the transcode block issue.
2. fix decode flush no frame issue.

Commit:
Merge branch 'master' of https://github.com/01org/ffmpeg_libyami into external-surface

Conflicts:
	libavcodec/libyami_dec.cpp

Commit:
Merge remote-tracking branch 'upstream/master' into external-surface

mypopydev commented on libavcodec/libyami_enc.h:
```diff
@@ -25,7 +25,7 @@
 
 
-#define ENCODE_QUEUE_SIZE 8
+#define ENCODE_QUEUE_SIZE 4
```
why change the defaults?

mypopydev commented on libavcodec/libyami_dec.cpp:
```diff
@@ -349,13 +360,19 @@ static int yami_dec_frame(AVCodecContext *avctx, void *data,
             av_log(avctx, AV_LOG_DEBUG, "getoutput() status=%d\n", status);
-        } while (!avpkt->data && !yami_image->output_frame && s->in_queue->size() > 0);
+            pthread_mutex_lock(&s->ctx_mutex);
+            if (!(!avpkt->data && !yami_image->output_frame && s->decode_status != DECODE_THREAD_EXIT)) {
```
OOP, clear the logic

mypopydev commented on libavcodec/libyami.cpp:
```diff
@@ -366,3 +366,49 @@ bool ff_vaapi_get_image(SharedPtr<VideoFrame>& frame, AVFrame *out)
+    return YAMI_SUCCESS;
+}
+
+YamiStatus ff_yami_free_surface (SurfaceAllocator* thiz, SurfaceAllocParams* params) {
```
coding style

mypopydev commented on libavcodec/libyami.cpp:
```diff
@@ -366,3 +366,49 @@ bool ff_vaapi_get_image(SharedPtr<VideoFrame>& frame, AVFrame *out)
     return true;
 }
+
+#define EXTRA_SIZE 14 //EXTRA_SIZE must great than DEC_QUE+ENC_QUE+DBP-19 or the thread will be block
```
put the DEC_QUE/ENC_QUE in the same file

Commit:
lavc : clean up source code.

1.move the queue max size into libyami.h.
2.modify the decode loop condition.

mypopydev commented on libavcodec/libyami.h:
```diff
@@ -52,4 +52,13 @@ bool ff_vaapi_load_image(SharedPtr<VideoFrame>& frame, AVFrame *in);
+YamiStatus ff_yami_free_surface (SurfaceAllocator* thiz, SurfaceAllocParams* params);
+void ff_yami_unref_surface (SurfaceAllocator* thiz);
+
+#define EXTRA_SIZE 14 //EXTRA_SIZE must great than DEC_QUE+ENC_QUE+DBP-19 or the thread will be block
```
more better #define EXTRA_SZIE  (DECODE_QUEUE_SIZE + ENCODE_QUEUE_SIZE + 2)

mypopydev commented on libavcodec/libyami_dec.cpp:
```diff
@@ -182,10 +125,7 @@ static void *ff_yami_decode_thread(void *arg)
         }
         s->decode_count_yami++;
-        pthread_mutex_lock(&s->in_mutex);
         s->in_queue->pop_front();
```
why remove the in queue lock ?

Commit:
lavc : libyami decoder clear logic.

1. in_queue->pop_front add mutex lock.
2. delete the no useful queue empty condition.
